lib/Analysis/Membar.cpp

```diff
@@ -229,6 +230,22 @@ void MembarAnalysis::update(Operation *op, BlockInfo *blockInfo,
   // read/write operations, and ending with a shared memory read, i.e., shared
   // memory write -> ... -> shared memory read.
   if (scratchBufferId != Allocation::InvalidBufferId) {
+    // Detect warp-synchronous convert-layout operations. These emit a
+    // warp-level barrier (warp.sync) rather than a CTA-wide barrier between
+    // the internal shared-memory write and read phases. For these ops, we must
+    // not globally clear pending dependencies.
+    bool isWarpSync = false;
+    if (auto cvt = dyn_cast<triton::gpu::ConvertLayoutOp>(op)) {
+      auto srcTy = cast<RankedTensorType>(cvt.getSrc().getType());
+      auto dstTy = cast<RankedTensorType>(cvt.getType());
+      auto srcLayout = triton::gpu::toLinearLayout(srcTy);
+      auto dstLayout = triton::gpu::toLinearLayout(dstTy);
+      auto kWarp = StringAttr::get(op->getContext(), "warp");
+      isWarpSync = srcLayout.invertAndCompose(dstLayout).isTrivialOver(kWarp) &&
+                   srcLayout.getFreeVariableMasks()[kWarp] == 0 &&
+                   dstLayout.getFreeVariableMasks()[kWarp] == 0;
+    }
```

Comment:
can we move that into a separate function and ideally share it with the lowering so that we make sure it is in sync

Reply:
Would it make sense to have an op interface that membar uses to ask the op if it synchronises, rather than hard-coding the expected behavior into membar?